rfcs/006 - Sync changes across databases.md

```diff
@@ -0,0 +1,84 @@
+
+We have leaned deeper into the infrastructure as data workflow by favoring two-way mode with transactions over IaC’s apply. This plays to our strengths and provides a compelling workflow to manage infrastructure. However, we need a good story to import or replicate changes from one database to the other since most people manage staging and production as separate cloud accounts.
+
+Two important requirements to keep in mind are:
+1. Being able to account for the fact that most of our joins are done using auto-incremented IDs which makes them different in every database.
+2. Users might want to exclude certain resources from the import workflow. In other words, there might be resources in staging that people don't want to replicate in production and vice versa. This requirement can be sidestepped initially, but whatever design we pick should be able to support this down the line.
+
+## Proposal
+
+There are too many edge cases around our data model and different kinds of staging->production arrangements companies might have (differences in capacity and secrets are a given, plus the two environments could be in different regions altogether) to try to reliably perform an import from one database to another automatically. The alternatives considering automatic workflows would fail or require manual intervention the majority of the time. The expectation for the workflow we set forth should then be that the user will still need to provide a decent amount of manual input.
+
+The IaSQL audit log table is timestamped and looking at the records from a given point in time for a staging database will give users most of the information they need to be able to replicate the changes in production. This proposal is grounded on that fact and seeks to make that workflow as smooth as possible by creating a feature that generates SQL from the audit log from a given point in time. The given point in time for this use case would be the last time staging and production were synced. The generated SQL will then be massaged and manually run by the user in production via a PostgreSQL client or a migration within a plain SQL migration system for teams that want to code review infrastructure changes. The IaSQL audit log is ordered chronologically and the SQL generated by a `iasql_get_sql_since()` function should not cause duplicate records or FK relation problems even if the expectation is that the generated SQL should be massaged.
```

Comment:
But if someone has never used IaSQL and already has a staging env and they want to replicate that in production, they will not be able to do it. This approach will work for someone with some IaSQL usage, right?

Reply:
And I'm not sure if I get the last sentence. The result from `iasql_get_sql_since()` won't be pretty similar to an export dump?

Reply:
> And I'm not sure if I get the last sentence. The result from `iasql_get_sql_since()` won't be pretty similar to an export dump?

an export dump generates the SQL for the entire database while the proposed function takes in a timestamp and generates SQL to replicate the changes from the provided timestamp until `now()`

Reply:
> But if someone has never used IaSQL and already has a staging env and they want to replicate that in production, they will not be able to do it. This approach will work for someone with some IaSQL usage, right?

This will work for someone that has two IaSQL databases: a staging and a production one

Reply:
> But if someone has never used IaSQL and already has a staging env and they want to replicate that in production, they will not be able to do it. This approach will work for someone with some IaSQL usage, right?

That's not really true? When you first install the modules the audit log still shows all of the already existing records as inserts, so if they really wanted to set the timestamp to before they created the database they can.

But the more common would be connecting staging and production to two databases, then making a change in staging, then getting the SQL that represents that change and minimally editing it to run in production.

Reply:
Oh, one thing we should probably do: for INSERT statements that are referencing an auto-generated ID, we should generate the SQL with a sub-query where we select on the `@cloudId` column's value, if available. It should be immediately clear to the user that they need to substitute that string with the production equivalent ID, and in some situations where the cloud ID is a user-provided string, it may not even need any changes.

Reply:
One problem i see about using audit logs is that the target account may be in a different state than the source one. An user may have done only 2 or 3 changes in _account 1_, but for some reason may have done other manual operations on _account 2_. If we base our sync in logs from certain changes, we are not going to properly reflect the status of account 1 into account 2.